src/ol/render/canvas/replay.js

```diff
@@ -360,7 +377,30 @@ ol.render.canvas.Replay.prototype.fill_ = function(context, rotation) {
   }
   context.fill();
   if (this.fillOrigin_) {
-    context.setTransform.apply(context, this.resetTransform_);
+    context.setTransform.apply(context, ol.render.canvas.resetTransform_);
+  }
+};
+
+
+/**
+ * @param {Array.<*>} declutterGroup Declutter group.
+ */
+ol.render.canvas.Replay.prototype.renderDeclutter_ = function(declutterGroup) {
+  if (declutterGroup && declutterGroup.length > 5) {
+    var groupCount = declutterGroup[4];
+    if (groupCount == 1 || groupCount == declutterGroup.length - 5) {
+      if (!this.declutterTree.collides(this.declutterTree.toBBox(declutterGroup))) {
+        this.declutterTree.insert(declutterGroup.slice(0, 4));
+        var drawImage = ol.render.canvas.drawImage;
+        for (var j = 5, jj = declutterGroup.length; j < jj; ++j) {
+          if (declutterGroup[j]) {
+            drawImage.apply(undefined, /** @type {Array} */ (declutterGroup[j]));
+          }
+        }
+      }
+      declutterGroup.length = 5;
```

Comment:
A lot of magic numbers, but I guess that is just how renderers / replays are right now.

Reply:
The structure is documented [here](https://github.com/openlayers/openlayers/pull/7328/files#diff-d6310f7e908b5b923bb2bc0abf9d7db2R43). I agree that these flat arrays are a bit hard to debug, but some believe they perform better than other structures.